File: leetcode_75/Binary_Search/python/875.py
from typing import List


class Solution:

    def check_answer(self, k, piles, h):
        s = 0
        for pile in piles:
            s += pile // k + (pile % k != 0)

        return True if h >= s else False

    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        left = 1
        right = 10 ** 9
        while left <= right:
            mid = (left + right) // 2
            if self.check_answer(mid, piles, h):
                right = mid - 1
            else:
                left = mid + 1

        return left

File: leetcode_75/Binary_Search/cpp/875.cpp
#include <vector>

using namespace std;

class Solution
{

    bool check_answer(long k, vector<int> &piles, int h)
    {
        long s = 0;
        for (auto pile : piles)
        {
            s += pile / k + (int)(pile % k != 0);
        }

        return h >= s ? true : false;
    }

public:
    int minEatingSpeed(vector<int> &piles, int h)
    {
        long left = 1;
        long right = 1000000000;
        long mid;
        while (left <= right)
        {
            mid = (left + right) / 2;
            if (this->check_answer(mid, piles, h))
            {
                right = mid - 1;
            }
            else
            {
                left = mid + 1;
            }
        }

        return left;
    }
};